sdjwt/src/main/kotlin/web5/security/SdJwt.kt

```diff
@@ -0,0 +1,366 @@
+    claimsVerifier.verify(issuerJwt.jwtClaimsSet, null)
+
+    if (verificationOptions.holderBindingOption == HolderBindingOption.VerifyHolderBinding) {
+      // If Holder Binding JWT is not provided, the Verifier MUST reject the Presentation.
+      require(keyBindingJwt != null) {
+        "Holder binding required, but holder binding JWT not found"
+      }
+
+      // Ensure that a signing algorithm was used that was deemed secure for the application. Refer to [RFC8725], Sections 3.1
+      // and 3.2 for details.
+      require(verificationOptions.supportedAlgorithms.contains(keyBindingJwt.header.algorithm)) {
+        "the algorithm of keyBindingJwt (${keyBindingJwt.header.algorithm}) is not part of the declared list of " +
+          "supported algorithms (${verificationOptions.supportedAlgorithms})"
+      }
+      // The none algorithm MUST NOT be accepted.
+      require(keyBindingJwt.header.algorithm != JWSAlgorithm.NONE) {
+        "algorithm in keyBindingJwt must not be `none`"
+      }
+
+      // Validate the signature over the Holder Binding JWT.
+      // Check that the Holder Binding JWT is valid using nbf, iat, and exp claims, if provided in the Holder Binding JWT.
+      // Determine that the Holder Binding JWT is bound to the current transaction and was created for this Verifier (replay
+      // protection). This is usually achieved by a nonce and aud field within the Holder Binding JWT.
+      val holderVerifier = keyBindingVerifier(verificationOptions)
+      require(keyBindingJwt.verify(holderVerifier)) {
+        throw SignatureException("Verifying the issuerJwt failed: ${keyBindingJwt.serialize()}")
+      }
+
+      val holderClaimsVerifier = DefaultJWTClaimsVerifier<SecurityContext>(
+        JWTClaimsSet.Builder()
+          .audience(verificationOptions.desiredAudience)
+          .claim("nonce", verificationOptions.desiredNonce)
+          .build(),
+        null
+      )
+      holderClaimsVerifier.verify(keyBindingJwt.jwtClaimsSet, null)
+    }
+  }
+
+  private fun issuerVerifier(verificationOptions: VerificationOptions): JWSVerifier {
+    val verifier = DefaultJWSVerifierFactory().createJWSVerifier(
+      issuerJwt.header,
+      jwkToKey(verificationOptions.issuerPublicJwk)
+    )
+    verifier.jcaContext.provider = BouncyCastleProviderSingleton.getInstance() as Provider
+    return verifier
+  }
+
+  private fun keyBindingVerifier(verificationOptions: VerificationOptions): JWSVerifier {
+    val verifier = DefaultJWSVerifierFactory().createJWSVerifier(
+      keyBindingJwt!!.header,
+      jwkToKey(verificationOptions.keyBindingPublicJwk!!)
+    )
+    verifier.jcaContext.provider = BouncyCastleProviderSingleton.getInstance() as Provider
+    return verifier
+  }
+
+  private fun jwkToKey(jwk: JWK): Key {
+    return when (jwk) {
+      is ECKey -> jwk.toPublicKey()
+      is OctetKeyPair -> jwk.toPublicKey()
+      is OctetSequenceKey -> jwk.toSecretKey()
```

Comment:
why is this secret but the others are public?

Reply:
Ooof, good catch. Removed this line. Should only be using asymmetric keys.